<img src="images/dask_horizontal.svg" align="right" width="30%">

# Bag: Parallel Lists for semi-structured data

Dask-bag excels in processing data that can be represented as a sequence of arbitrary inputs. We'll refer to this as "messy" data, because it can contain complex nested structures, missing fields, mixtures of data types, etc. The *functional* programming style fits very nicely with standard Python iteration, such as can be found in the `itertools` module.

Messy data is often encountered at the beginning of data processing pipelines when large volumes of raw data are first consumed. The initial set of data might be JSON, CSV, XML, or any other format that does not enforce strict structure and datatypes.
For this reason, the initial data massaging and processing is often done with Python `list`s, `dict`s, and `set`s.

These core data structures are optimized for general-purpose storage and processing.  Adding streaming computation with iterators/generator expressions or libraries like `itertools` or [`toolz`](https://toolz.readthedocs.io/en/latest/) let us process large volumes in a small space.  If we combine this with parallel processing then we can churn through a fair amount of data.

Dask.bag is a high level Dask collection to automate common workloads of this form.  In a nutshell

    dask.bag = map, filter, toolz + parallel execution
    
**Related Documentation**

* [Bag documentation](https://docs.dask.org/en/latest/bag.html)
* [Bag screencast](https://youtu.be/-qIiJ1XtSv0)
* [Bag API](https://docs.dask.org/en/latest/bag-api.html)
* [Bag examples](https://examples.dask.org/bag.html)

## Create data

In [1]:
%run prep.py -d accounts

## Setup

Again, we'll use the distributed scheduler. Schedulers will be explained in depth [later](05_distributed.ipynb).

In [2]:
from dask.distributed import Client

client = Client(n_workers=4)

## Creation

You can create a `Bag` from a Python sequence, from files, from data on S3, etc.
We demonstrate using `.take()` to show elements of the data. (Doing `.take(1)` results in a tuple with one element)

Note that the data are partitioned into blocks, and there are many items per block. In the first example, the two partitions contain five elements each, and in the following two, each file is partitioned into one or more bytes blocks.

In [3]:
# each element is an integer
import dask.bag as db
b = db.from_sequence([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], npartitions=2)
b.take(3)

(1, 2, 3)

In [4]:
# each element is a text file, where each line is a JSON object
# note that the compression is handled automatically
import os
b = db.read_text(os.path.join('data', 'accounts.*.json.gz'))
b.take(1)

('{"id": 0, "name": "Ursula", "transactions": [{"transaction-id": 16, "amount": 131}, {"transaction-id": 164, "amount": 168}, {"transaction-id": 234, "amount": 121}, {"transaction-id": 237, "amount": 177}, {"transaction-id": 299, "amount": 126}, {"transaction-id": 325, "amount": 101}, {"transaction-id": 336, "amount": 141}, {"transaction-id": 763, "amount": 175}, {"transaction-id": 882, "amount": 131}, {"transaction-id": 1133, "amount": 109}, {"transaction-id": 1159, "amount": 136}, {"transaction-id": 1238, "amount": 132}, {"transaction-id": 1314, "amount": 112}, {"transaction-id": 1380, "amount": 143}, {"transaction-id": 1415, "amount": 126}, {"transaction-id": 1706, "amount": 118}, {"transaction-id": 2132, "amount": 148}, {"transaction-id": 2139, "amount": 113}, {"transaction-id": 2153, "amount": 158}, {"transaction-id": 2267, "amount": 130}, {"transaction-id": 2396, "amount": 112}, {"transaction-id": 2427, "amount": 122}, {"transaction-id": 2430, "amount": 123}, {"transaction-id": 2

In [5]:
# Requires `s3fs` library
# each partition is a remote CSV text file
b = db.read_text('s3://dask-data/nyc-taxi/2015/yellow_tripdata_2015-01.csv',
                 storage_options={'anon': True})
b.take(1)

('VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,pickup_longitude,pickup_latitude,RateCodeID,store_and_fwd_flag,dropoff_longitude,dropoff_latitude,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount\n',)

## Manipulation

`Bag` objects hold the standard functional API found in projects like the Python standard library, `toolz`, or `pyspark`, including `map`, `filter`, `groupby`, etc..

Operations on `Bag` objects create new bags.  Call the `.compute()` method to trigger execution, as we saw for `Delayed` objects.  

In [6]:
def is_even(n):
    return n % 2 == 0

b = db.from_sequence([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
c = b.filter(is_even).map(lambda x: x ** 2)
c

dask.bag<lambda, npartitions=10>

In [7]:
# blocking form: wait for completion (which is very fast in this case)
c.compute()

[4, 16, 36, 64, 100]

### Example: Accounts JSON data

We've created a fake dataset of gzipped JSON data in your data directory.  This is like the example used in the `DataFrame` example we will see later, except that it has bundled up all of the entires for each individual `id` into a single record.  This is similar to data that you might collect off of a document store database or a web API.

Each line is a JSON encoded dictionary with the following keys

*  id: Unique identifier of the customer
*  name: Name of the customer
*  transactions: List of `transaction-id`, `amount` pairs, one for each transaction for the customer in that file

In [8]:
filename = os.path.join('data', 'accounts.*.json.gz')
lines = db.read_text(filename)
lines.take(3)

('{"id": 0, "name": "Ursula", "transactions": [{"transaction-id": 16, "amount": 131}, {"transaction-id": 164, "amount": 168}, {"transaction-id": 234, "amount": 121}, {"transaction-id": 237, "amount": 177}, {"transaction-id": 299, "amount": 126}, {"transaction-id": 325, "amount": 101}, {"transaction-id": 336, "amount": 141}, {"transaction-id": 763, "amount": 175}, {"transaction-id": 882, "amount": 131}, {"transaction-id": 1133, "amount": 109}, {"transaction-id": 1159, "amount": 136}, {"transaction-id": 1238, "amount": 132}, {"transaction-id": 1314, "amount": 112}, {"transaction-id": 1380, "amount": 143}, {"transaction-id": 1415, "amount": 126}, {"transaction-id": 1706, "amount": 118}, {"transaction-id": 2132, "amount": 148}, {"transaction-id": 2139, "amount": 113}, {"transaction-id": 2153, "amount": 158}, {"transaction-id": 2267, "amount": 130}, {"transaction-id": 2396, "amount": 112}, {"transaction-id": 2427, "amount": 122}, {"transaction-id": 2430, "amount": 123}, {"transaction-id": 2

Our data comes out of the file as lines of text. Notice that file decompression happened automatically. We can make this data look more reasonable by mapping the `json.loads` function onto our bag.

In [9]:
import json
js = lines.map(json.loads)
# take: inspect first few elements
js.take(3)

({'id': 0,
  'name': 'Ursula',
  'transactions': [{'transaction-id': 16, 'amount': 131},
   {'transaction-id': 164, 'amount': 168},
   {'transaction-id': 234, 'amount': 121},
   {'transaction-id': 237, 'amount': 177},
   {'transaction-id': 299, 'amount': 126},
   {'transaction-id': 325, 'amount': 101},
   {'transaction-id': 336, 'amount': 141},
   {'transaction-id': 763, 'amount': 175},
   {'transaction-id': 882, 'amount': 131},
   {'transaction-id': 1133, 'amount': 109},
   {'transaction-id': 1159, 'amount': 136},
   {'transaction-id': 1238, 'amount': 132},
   {'transaction-id': 1314, 'amount': 112},
   {'transaction-id': 1380, 'amount': 143},
   {'transaction-id': 1415, 'amount': 126},
   {'transaction-id': 1706, 'amount': 118},
   {'transaction-id': 2132, 'amount': 148},
   {'transaction-id': 2139, 'amount': 113},
   {'transaction-id': 2153, 'amount': 158},
   {'transaction-id': 2267, 'amount': 130},
   {'transaction-id': 2396, 'amount': 112},
   {'transaction-id': 2427, 'amount': 1

### Basic Queries

Once we parse our JSON data into proper Python objects (`dict`s, `list`s, etc.) we can perform more interesting queries by creating small Python functions to run on our data.

In [10]:
# filter: keep only some elements of the sequence
js.filter(lambda record: record['name'] == 'Alice').take(5)

({'id': 15,
  'name': 'Alice',
  'transactions': [{'transaction-id': 111, 'amount': 398},
   {'transaction-id': 1040, 'amount': 405},
   {'transaction-id': 1121, 'amount': 500},
   {'transaction-id': 1151, 'amount': 468},
   {'transaction-id': 1457, 'amount': 388},
   {'transaction-id': 1611, 'amount': 450},
   {'transaction-id': 1675, 'amount': 396},
   {'transaction-id': 1781, 'amount': 516},
   {'transaction-id': 1815, 'amount': 415},
   {'transaction-id': 1845, 'amount': 387},
   {'transaction-id': 1861, 'amount': 503},
   {'transaction-id': 1864, 'amount': 475},
   {'transaction-id': 1900, 'amount': 373},
   {'transaction-id': 2054, 'amount': 489},
   {'transaction-id': 2161, 'amount': 427},
   {'transaction-id': 2272, 'amount': 471},
   {'transaction-id': 2880, 'amount': 453},
   {'transaction-id': 3157, 'amount': 378},
   {'transaction-id': 3248, 'amount': 472},
   {'transaction-id': 3340, 'amount': 531},
   {'transaction-id': 3396, 'amount': 548},
   {'transaction-id': 3643, 'a

In [11]:
def count_transactions(d):
    return {'name': d['name'], 'count': len(d['transactions'])}

# map: apply a function to each element
(js.filter(lambda record: record['name'] == 'Alice')
   .map(count_transactions)
   .take(5))

({'name': 'Alice', 'count': 54},
 {'name': 'Alice', 'count': 30},
 {'name': 'Alice', 'count': 5},
 {'name': 'Alice', 'count': 23},
 {'name': 'Alice', 'count': 88})

In [12]:
# pluck: select a field, as from a dictionary, element[field]
(js.filter(lambda record: record['name'] == 'Alice')
   .map(count_transactions)
   .pluck('count')
   .take(5))

(54, 30, 5, 23, 88)

In [13]:
# Average number of transactions for all of the Alice entries
(js.filter(lambda record: record['name'] == 'Alice')
   .map(count_transactions)
   .pluck('count')
   .mean()
   .compute())

39.749640287769786

### Use `flatten` to de-nest

In the example below we see the use of `.flatten()` to flatten results.  We compute the average amount for all transactions for all Alices.

In [14]:
js.filter(lambda record: record['name'] == 'Alice').pluck('transactions').take(3)

([{'transaction-id': 111, 'amount': 398},
  {'transaction-id': 1040, 'amount': 405},
  {'transaction-id': 1121, 'amount': 500},
  {'transaction-id': 1151, 'amount': 468},
  {'transaction-id': 1457, 'amount': 388},
  {'transaction-id': 1611, 'amount': 450},
  {'transaction-id': 1675, 'amount': 396},
  {'transaction-id': 1781, 'amount': 516},
  {'transaction-id': 1815, 'amount': 415},
  {'transaction-id': 1845, 'amount': 387},
  {'transaction-id': 1861, 'amount': 503},
  {'transaction-id': 1864, 'amount': 475},
  {'transaction-id': 1900, 'amount': 373},
  {'transaction-id': 2054, 'amount': 489},
  {'transaction-id': 2161, 'amount': 427},
  {'transaction-id': 2272, 'amount': 471},
  {'transaction-id': 2880, 'amount': 453},
  {'transaction-id': 3157, 'amount': 378},
  {'transaction-id': 3248, 'amount': 472},
  {'transaction-id': 3340, 'amount': 531},
  {'transaction-id': 3396, 'amount': 548},
  {'transaction-id': 3643, 'amount': 440},
  {'transaction-id': 3679, 'amount': 370},
  {'transact

In [15]:
(js.filter(lambda record: record['name'] == 'Alice')
   .pluck('transactions')
   .flatten()
   .take(3))

({'transaction-id': 111, 'amount': 398},
 {'transaction-id': 1040, 'amount': 405},
 {'transaction-id': 1121, 'amount': 500})

In [16]:
(js.filter(lambda record: record['name'] == 'Alice')
   .pluck('transactions')
   .flatten()
   .pluck('amount')
   .take(3))

(398, 405, 500)

In [17]:
(js.filter(lambda record: record['name'] == 'Alice')
   .pluck('transactions')
   .flatten()
   .pluck('amount')
   .mean()
   .compute())

1881.7921161224933

### Groupby and Foldby

Often we want to group data by some function or key.  We can do this either with the `.groupby` method, which is straightforward but forces a full shuffle of the data (expensive) or with the harder-to-use but faster `.foldby` method, which does a streaming combined groupby and reduction.

*  `groupby`:  Shuffles data so that all items with the same key are in the same key-value pair
*  `foldby`:  Walks through the data accumulating a result per key

*Note: the full groupby is particularly bad. In actual workloads you would do well to use `foldby` or switch to `DataFrame`s if possible.*

### `groupby`

Groupby collects items in your collection so that all items with the same value under some function are collected together into a key-value pair.

In [18]:
b = db.from_sequence(['Alice', 'Bob', 'Charlie', 'Dan', 'Edith', 'Frank'])
b.groupby(len).compute()  # names grouped by length

[(7, ['Charlie']), (3, ['Bob', 'Dan']), (5, ['Alice', 'Edith', 'Frank'])]

In [19]:
b = db.from_sequence(list(range(10)))
b.groupby(lambda x: x % 2).compute()

[(0, [0, 2, 4, 6, 8]), (1, [1, 3, 5, 7, 9])]

In [20]:
b.groupby(lambda x: x % 2).starmap(lambda k, v: (k, max(v))).compute()

[(0, 8), (1, 9)]

### `foldby`

Foldby can be quite odd at first.  It is similar to the following functions from other libraries:

*  [`toolz.reduceby`](http://toolz.readthedocs.io/en/latest/streaming-analytics.html#streaming-split-apply-combine)
*  [`pyspark.RDD.combineByKey`](http://abshinn.github.io/python/apache-spark/2014/10/11/using-combinebykey-in-apache-spark/)

When using `foldby` you provide 

1.  A key function on which to group elements
2.  A binary operator such as you would pass to `reduce` that you use to perform reduction per each group
3.  A combine binary operator that can combine the results of two `reduce` calls on different parts of your dataset.

Your reduction must be associative.  It will happen in parallel in each of the partitions of your dataset.  Then all of these intermediate results will be combined by the `combine` binary operator.

In [21]:
is_even = lambda x: x % 2
b.foldby(is_even, binop=max, combine=max).compute()

[(0, 8), (1, 9)]

### Example with account data

We find the number of people with the same name.

In [22]:
%%time
# Warning, this one takes a while...
result = js.groupby(lambda item: item['name']).starmap(lambda k, v: (k, len(v))).compute()
print(sorted(result))

[('Alice', 168), ('Alice', 168), ('Alice', 179), ('Alice', 180), ('Bob', 36), ('Bob', 39), ('Bob', 75), ('Charlie', 118), ('Charlie', 119), ('Charlie', 128), ('Charlie', 129), ('Dan', 144), ('Dan', 144), ('Dan', 155), ('Dan', 156), ('Edith', 60), ('Edith', 60), ('Edith', 65), ('Edith', 65), ('Frank', 108), ('Frank', 108), ('Frank', 117), ('Frank', 117), ('George', 157), ('George', 158), ('George', 171), ('George', 172), ('Hannah', 81), ('Hannah', 82), ('Hannah', 88), ('Hannah', 88), ('Ingrid', 159), ('Ingrid', 165), ('Ingrid', 176), ('Ingrid', 178), ('Jerry', 72), ('Jerry', 72), ('Jerry', 78), ('Jerry', 78), ('Kevin', 108), ('Kevin', 108), ('Kevin', 117), ('Kevin', 117), ('Laura', 77), ('Laura', 78), ('Laura', 82), ('Laura', 83), ('Michael', 116), ('Michael', 118), ('Michael', 128), ('Michael', 129), ('Norbert', 116), ('Norbert', 116), ('Norbert', 123), ('Norbert', 126), ('Oliver', 39), ('Oliver', 39), ('Oliver', 72), ('Patricia', 139), ('Patricia', 142), ('Patricia', 149), ('Patricia'

In [23]:
%%time
# This one is comparatively fast and produces the same result.
from operator import add
def incr(tot, _):
    return tot+1

result = js.foldby(key='name', 
                   binop=incr, 
                   initial=0, 
                   combine=add, 
                   combine_initial=0).compute()
print(sorted(result))

[('Alice', 695), ('Bob', 150), ('Charlie', 494), ('Dan', 599), ('Edith', 250), ('Frank', 450), ('George', 658), ('Hannah', 339), ('Ingrid', 678), ('Jerry', 300), ('Kevin', 450), ('Laura', 320), ('Michael', 491), ('Norbert', 481), ('Oliver', 150), ('Patricia', 581), ('Quinn', 296), ('Ray', 350), ('Sarah', 397), ('Tim', 400), ('Ursula', 750), ('Victor', 482), ('Wendy', 693), ('Xavier', 585), ('Yvonne', 600), ('Zelda', 586)]
CPU times: user 197 ms, sys: 3.35 ms, total: 201 ms
Wall time: 645 ms


### Exercise: compute total amount per name

We want to groupby (or foldby) the `name` key, then add up the all of the amounts for each name.

Steps

1.  Create a small function that, given a dictionary like 

        {'name': 'Alice', 'transactions': [{'amount': 1, 'id': 123}, {'amount': 2, 'id': 456}]}
        
    produces the sum of the amounts, e.g. `3`
    
2.  Slightly change the binary operator of the `foldby` example above so that the binary operator doesn't count the number of entries, but instead accumulates the sum of the amounts.

In [24]:
# Your code here...

## DataFrames

For the same reasons that Pandas is often faster than pure Python, `dask.dataframe` can be faster than `dask.bag`.  We will work more with DataFrames later, but from for the bag point of view, they are frequently the end-point of the "messy" part of data ingestion—once the data can be made into a data-frame, then complex split-apply-combine logic will become much more straight-forward and efficient.

You can transform a bag with a simple tuple or flat dictionary structure into a `dask.dataframe` with the `to_dataframe` method.

In [25]:
df1 = js.to_dataframe()
df1.head()

id      name                                       transactions
0   0    Ursula  [{'transaction-id': 16, 'amount': 131}, {'tran...
1   1       Dan  [{'transaction-id': 813, 'amount': 1486}, {'tr...
2   2  Patricia  [{'transaction-id': 531, 'amount': 1319}, {'tr...
3   3    Victor  [{'transaction-id': 1266, 'amount': 126}, {'tr...
4   4     Kevin  [{'transaction-id': 288, 'amount': -650}, {'tr...

This now looks like a well-defined DataFrame, and we can apply Pandas-like computations to it efficiently.

Using a Dask DataFrame, how long does it take to do our prior computation of numbers of people with the same name?  It turns out that `dask.dataframe.groupby()` beats `dask.bag.groupby()` more than an order of magnitude; but it still cannot match `dask.bag.foldby()` for this case.

In [26]:
%time df1.groupby('name').id.count().compute().head()

CPU times: user 244 ms, sys: 17.2 ms, total: 262 ms
Wall time: 1.85 s


name
Alice      695
Bob        150
Charlie    494
Dan        599
Edith      250
Name: id, dtype: int64

### Denormalization

This DataFrame format is less-than-optimal because the `transactions` column is filled with nested data so Pandas has to revert to `object` dtype, which is quite slow in Pandas.  Ideally we want to transform to a dataframe only after we have flattened our data so that each record is a single `int`, `string`, `float`, etc..

In [27]:
def denormalize(record):
    # returns a list for every nested item, each transaction of each person
    return [{'id': record['id'], 
             'name': record['name'], 
             'amount': transaction['amount'], 
             'transaction-id': transaction['transaction-id']}
            for transaction in record['transactions']]

transactions = js.map(denormalize).flatten()
transactions.take(3)

({'id': 0, 'name': 'Ursula', 'amount': 131, 'transaction-id': 16},
 {'id': 0, 'name': 'Ursula', 'amount': 168, 'transaction-id': 164},
 {'id': 0, 'name': 'Ursula', 'amount': 121, 'transaction-id': 234})

In [28]:
df = transactions.to_dataframe()
df.head()

id    name  amount  transaction-id
0   0  Ursula     131              16
1   0  Ursula     168             164
2   0  Ursula     121             234
3   0  Ursula     177             237
4   0  Ursula     126             299

In [29]:
%%time
# number of transactions per name
# note that the time here includes the data load and ingestion
df.groupby('name')['transaction-id'].count().compute()

CPU times: user 247 ms, sys: 15.5 ms, total: 263 ms
Wall time: 1.56 s


name
Alice       27626
Bob          5187
Charlie     12249
Dan         32516
Edith        5778
Frank       22268
George      21735
Hannah      17186
Ingrid      18984
Jerry       11212
Kevin       17400
Laura       14477
Michael     32848
Norbert     15071
Oliver      11797
Patricia    19442
Quinn       10694
Ray         11705
Sarah       19191
Tim         19510
Ursula      29982
Victor      19294
Wendy       22265
Xavier      27943
Yvonne      26448
Zelda       27192
Name: transaction-id, dtype: int64

## Limitations

Bags provide very general computation (any Python function.)  This generality
comes at cost.  Bags have the following known limitations

1.  Bag operations tend to be slower than array/dataframe computations in the
    same way that Python tends to be slower than NumPy/Pandas
2.  ``Bag.groupby`` is slow.  You should try to use ``Bag.foldby`` if possible.
    Using ``Bag.foldby`` requires more thought. Even better, consider creating
    a normalised dataframe.

## Learn More

* [Bag documentation](https://docs.dask.org/en/latest/bag.html)
* [Bag screencast](https://youtu.be/-qIiJ1XtSv0)
* [Bag API](https://docs.dask.org/en/latest/bag-api.html)
* [Bag examples](https://examples.dask.org/bag.html)

## Shutdown

In [30]:
client.shutdown()